File: cplusplus/src/sign.cpp
#include "sign.h"

#include "error.h"
#include "sizes.h"

namespace Autograph {

SignFunction createSign(const std::vector<unsigned char> identityPrivateKey) {
  SignFunction sign = [identityPrivateKey](std::vector<unsigned char> subject) {
    std::vector<unsigned char> signature(SIGNATURE_SIZE);
    bool success =
        autograph_sign_subject(signature.data(), identityPrivateKey.data(),
                               subject.data(), subject.size()) == 0;
    if (!success) {
      throw Error(Error::Signing);
    }
    return signature;
  };
  return sign;
}

}  // namespace Autograph

File: swift/Sources/Autograph/Sign.swift
import Clibautograph
import Foundation

public typealias SignFunction = ([UInt8]) throws -> [UInt8]

public func createSign(identityPrivateKey: [UInt8]) -> SignFunction {
  let sign: SignFunction = { [identityPrivateKey] subject in
    var signature = createSignatureBytes()
    let success =
      autograph_sign_subject(
        &signature,
        identityPrivateKey,
        subject,
        UInt32(subject.count)
      ) == 0
    if !success {
      throw AutographError.signing
    }
    return signature
  }
  return sign
}
